contacts/views.py

```diff
@@ -0,0 +1,151 @@
+from django.contrib.auth.decorators import login_required
+from django.core.exceptions import ValidationError
+from django.shortcuts import render, get_object_or_404
+from django.contrib.auth import authenticate, login, logout
+from django.http import HttpResponse, HttpResponseNotFound
+from django.http import HttpResponseRedirect
+
+from contacts.forms import NewContactForm
+from contacts.models import Contact, Email, Number
+from django.contrib.auth.models import User
+from django.views.decorators.http import require_http_methods
+from django.contrib.auth.decorators import login_required
+
+
+@login_required
+@require_http_methods(["POST", "GET"])
+def add_contact(request):
+    if request.method == "GET":
+        form = NewContactForm()
+        return render(request, 'contacts/add_contact.html', {'form': form})
+    else:
+        form = NewContactForm(request.POST)
+        if not form.is_valid():
```

Comment:
Good job with the form validation. However, when a user enters a mistake it causes your app to show an error and the user loses all the entered data. Please preserve this state somehow while showing the user the errors as well.

Reply:
Also, try to do this without JavaScript.